javatests/arcs/core/analysis/ExpressionClaimDeducerTest.kt

```diff
@@ -0,0 +1,348 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.core.analysis
+
+import arcs.core.data.expression.PaxelParser
+import com.google.common.truth.Truth.assertThat
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+
+@RunWith(JUnit4::class)
+class ExpressionClaimDeducerTest {
+
+  @Test
+  fun literal_number() {
+    val expr = PaxelParser.parse("5")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(Deduction.Path())
+  }
+
+  @Test
+  fun literal_text() {
+    val expr = PaxelParser.parse("'hello'")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(Deduction.Path())
+  }
+
+  @Test
+  fun literal_boolean() {
+    val expr = PaxelParser.parse("false")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(Deduction.Path())
+  }
+
+  @Test
+  fun literal_null() {
+    val expr = PaxelParser.parse("null")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(Deduction.Path())
+  }
+
+  @Test
+  fun single_variable() {
+    val expr = PaxelParser.parse("x")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Equal(
+        Deduction.Path("x")
+      )
+    )
+  }
+
+  @Test
+  fun variable_field_access() {
+    val expr = PaxelParser.parse("x.foo")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Equal(
+        Deduction.Path("x", "foo")
+      )
+    )
+  }
+
+  @Test
+  fun variable_field_access_nested() {
+    val expr = PaxelParser.parse("x.foo.bar")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Equal(
+        Deduction.Path("x", "foo", "bar")
+      )
+    )
+  }
+
+  @Test
+  fun variable_field_access_binop() {
+    val expr = PaxelParser.parse("x.foo.bar + y.foo.bar.baz + z.baz.bar")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Derive(
+        Deduction.Paths(
```

Comment:
I am pondering whether you need the nesting of `Path` inside `Deduction.Equal` and nesting of `Paths` inside `Deduction.Derive`. It appears like Paths are only used as a member of derive and Path is only ever a member of equal.

Reply:
What about the hypothetical `InfluencedBy` claim? Will we want `Paths` and `Path` then? What if we add another claim type later?

Reply:
@bgogul What's your perspective on the issue?

Reply:
Thought about it a bit more: Another way of looking at your point might be -- Can't we always calculate the claim type from the nested structure? `Paths` always implies a `Derive`, whereas `Path` does not. So, why not get rid of the Claim subtypes all together?

Reply:
If we ever need `InfluencedBy` claim, it would be similar to `Derive` and also hold a list of paths. But what could hypothetically come someday should not shape the design of the code you have today if you can refactor easily in the future.